src/recon_buildblock/PLSPrior.cxx

```diff
@@ -0,0 +1,607 @@
+//
+//
+/*
+    Copyright (C) 2018 University of Leeds and University College of London
+                  2000- 2011, Hammersmith Imanet Ltd
+    This file is part of STIR.
+
+    This file is free software; you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 2.1 of the License, or
+    (at your option) any later version.
+
+    This file is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License for more details.
+
+    See STIR/LICENSE.txt for details
+*/
+/*!
+  \file
+  \ingroup priors
+  \brief  implementation of the stir::PLSPrior class
+
+  \author Daniel Deidda
+  \author Tsai Yu-Jung
+
+*/
+
+#include "stir/recon_buildblock/PLSPrior.h"
+#include "stir/Succeeded.h"
+#include "stir/DiscretisedDensityOnCartesianGrid.h"
+#include "stir/IndexRange3D.h"
+#include "stir/IO///write_to_file.h"
+#include "stir/IO/read_from_file.h"
+#include "stir/is_null_ptr.h"
+#include "stir/info.h"
+#include <algorithm>
+using std::min;
+using std::max;
+
+/* Pretty horrible code because we don't have an iterator of neigbhourhoods yet
+ */
+
+START_NAMESPACE_STIR
+
+template <typename elemT>
+void
+PLSPrior<elemT>::initialise_keymap()
+{
+  base_type::initialise_keymap();
+  this->parser.add_start_key("PLS Prior Parameters");
+  this->parser.add_key("only 2D", &only_2D);
+  this->parser.add_key("eta", &eta);
+  this->parser.add_key("alpha", &alpha);
+  this->parser.add_key("kappa filename", &kappa_filename);
+  this->parser.add_key("anatomical_filename", &anatomical_filename);
+  this->parser.add_key("gradient filename prefix", &gradient_filename_prefix);
+  this->parser.add_stop_key("END PLS Prior Parameters");
+}
+
+
+template <typename elemT>
+Succeeded
+PLSPrior<elemT>::set_up ()
+{
```

Comment:
need a check
```c++
if (is_null_ptr( this->anatomical_sptr))
{
   error("PLS prior needs an anatomical image");
  return Succeeded::no;
}
```
I know you'd want to set it to zero, but we don't know the appriopriate size yet. Let's leave this for later (add a `\todo` comment in the doxygen of the class)

Reply:
Ok, Shall I then remove the part where I say that if no anatomical image is added the prior becomes smoothed TV?

Reply:
you could say "if you provide a uniform anatomical image..."